kong/tools/mime_type.lua

```diff
@@ -0,0 +1,111 @@
+local lpeg = require "lpeg"
+local clear_tab = require "table.clear"
+
+local P, S, R, C = lpeg.P, lpeg.S, lpeg.R, lpeg.C
+local ipairs = ipairs
+local lower = string.lower
+local EMPTY = {}
+
+--[[
+RFC2045(https://www.ietf.org/rfc/rfc2045.txt)
+
+media-type     = type "/" subtype *(";" parameter )
+parameter      = attribute "=" value
+attribute      = token
+value          = token | quoted-string
+quoted-string  = ( <"> *(qdtext | quoted-pair ) <"> )
+qdtext         = <any TEXT except <">>
+quoted-pair    = "\" CHAR
+type           = token
+subtype        = token
+token          = 1*<any CHAR except CTLs or separators>
+CHAR           = <any US-ASCII character (octets 0 - 127)>
+separators     = "(" | ")" | "<" | ">" | "@"
+               | "," | ";" | ":" | "\" | <">
+               | "/" | "[" | "]" | "?" | "="
+               | "{" | "}" | SP | HT
+CTL            = <any US-ASCII ctl chr (0-31) and DEL (127)>
+]]--
+
+local CTL = R"\0\31" + P"\127"
+local CHAR = R"\0\127"
+local quote = P'"'
+local separators = S"()<>@,;:\\\"/[]?={} \t"
+local token = (CHAR - CTL - separators)^1
+local spacing = (S" \t")^0
+
+local qdtext = P(1) - CTL - quote
+local quoted_pair = P"\\" * CHAR
+local quoted_string = quote * C((qdtext + quoted_pair)^0) * quote
+
+local attribute = C(token)
+local value = C(token) + quoted_string
+local parameter = attribute * P"=" * value
+local parameters = (spacing * P";" * spacing * parameter)^0
+local types = C(token) * P"/" * C(token) + C"*"
+
+local function format_types(...)
+  local args = {...}
+  if #args == 1 and args[1] == "*" then
+    return "*", "*"
+  end
+  return ...
+end
+
+-- cached vars
+local param_key_ignorecase
+local ignorecase_params = {}
```

Comment:
A module-level table like this will be shared between requests and is likely to cause bugs. Please consider refactoring this table to a function-scope variable.

Reply:
This was designed to be able to read the options parameter passing in `parse_mime_type`

A worker in general can process a single request at the same time. I don't see data race conditions in here.